en/_posts/2024-05-23-understanding-github-artifact-attestations.md

```diff
@@ -0,0 +1,224 @@
+contexts. These include the inputs into a build and could potentially be used
+for script injection attacks. While they can’t be used to attack the provenance
+itself, they could be used to attack the build without leaving any trace in the
+provenance.
+
+This means that while GitHub Artifact Attestations can be used with policy
+engines/tools like Rego, Cue, and OPA, these policies can’t be evaluated on the
+build inputs. For example, to verify that the build had no inputs etc. in order
+to prevent these kinds of attacks.
+
+This architecture also relies heavily on Sigstore’s Fulcio Certificate
+Authority to map the OID claims. This means that it’s hard to use separate
+private PKI, such as static keys, to sign and verify provenance. While this
+isn’t likely a problem for Open Source projects, some enterprises have these
+kinds of requirements.
+
+## Achieving SLSA Build L3 with GitHub Artifact Attestations
+
+In the documentation [SLSA Build L3 is described as
+achievable](https://github.blog/2024-05-02-introducing-artifact-attestations-now-in-public-beta/#an-effortless-user-experience)
+but it focuses on SLSA Build L2(+) and L3 is left as an exercise for the
+reader. So what would it take to achieve SLSA Build L3 with GitHub Artifact
+Attestations?
+
+The solution is fairly somewhat so I will just summarize the requirements here
+and leave the details for a later blog post. In order to achieve L3 you, at a
+minimum, would roughly need to do the following things:
+
+- The artifact build would need to be done in a separate job from the call to
+  `attest-build-provenance`. This ensures that the build and provenance signing
+  happen on [separate virtual
+  machines](https://docs.github.com/en/actions/learn-github-actions/understanding-github-actions#the-components-of-github-actions)
+  and that the artifact build steps will not have access to the signing
+  material. This job should **not** have `attestations: write` permissions.
+- The `attest-build-provenance` action is passed a path to the artifact. Jobs
+  execute on separate VMs so the artifact itself would need to be passed from
+  the build job to the attestation job. This will need to be done by uploading
+  the artifact file as a [workflow
+  artifact](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts)
+  from the first job and downloading it in the second job. However, this
+  artifact could be overwritten by concurrently running build steps. So you
+  will need to take a checksum (sha256 etc.) and pass it from the build job to
+  the provenance signing job using job inputs/outputs, and verify it after
+  download.
+
+## Conclusion
+
+Trust in GitHub’s Artifact Attestations trust really lies in the Sigstore
+certificate and its OID claims. The signed SLSA provenance is really only nice
+to have. GitHub ensures that the JSON printed by the `gh` client’s `--format
+json` option is trustworthy simply by omitting any values from the SLSA
```

Comment:
why do they allow non-trusted values in the intoto payload? They should not imo.

Reply:
They don't afaik, though I haven't tested adding extra fields to the predicate and seeing if the attestation store accepts it.

I really meant that they are validating that the fields are the same, and that's why it's trustworthy but only just so. Maybe what I wrote is just too opaque.